stan/math/rev/mat/functor/kinsol_data.hpp

```diff
@@ -0,0 +1,146 @@
+
+    std::move(jacobian_x.begin(), jacobian_x.end(), SM_DATA_D(J));
+
+    return 0;
+  }
+};
+
+/**
+ * KINSOL algebraic system data holder.
+ * Based on cvodes_ode_data.
+ *
+ * @tparam F1 functor type for system function.
+ * @tparam F2 functor type for jacobian function. Default is 0.
+ *         If 0, use rev mode autodiff to compute the Jacobian.
+ */
+template <typename F1, typename F2>
+class kinsol_system_data {
+  const F1& f_;
+  const F2& J_f_;
+  const Eigen::VectorXd& x_;
+  const Eigen::VectorXd& y_;
+  const size_t N_;
+  const std::vector<double>& dat_;
+  const std::vector<int>& dat_int_;
+  std::ostream* msgs_;
+
+  typedef kinsol_system_data<F1, F2> system_data;
+
+ public:
+  N_Vector nv_x_;
+  SUNMatrix J_;
+  SUNLinearSolver LS_;
+
+  /* Constructor */
+  kinsol_system_data(const F1& f, const F2& J_f, const Eigen::VectorXd& x,
+                     const Eigen::VectorXd& y, const std::vector<double>& dat,
+                     const std::vector<int>& dat_int, std::ostream* msgs)
+      : f_(f),
+        J_f_(J_f),
+        x_(x),
+        y_(y),
+        dat_(dat),
+        dat_int_(dat_int),
+        msgs_(msgs),
+        N_(x.size()),
+        nv_x_(N_VMake_Serial(N_, &to_array_1d(x_)[0])),
+        J_(SUNDenseMatrix(N_, N_)),
+        LS_(SUNLinSol_Dense(nv_x_, J_)) {}
+
+  ~kinsol_system_data() {
+    N_VDestroy_Serial(nv_x_);
+    SUNLinSolFree(LS_);
+    SUNMatDestroy(J_);
+  }
+
+  /* Implements the user-defined function passed to KINSOL. */
+  static int kinsol_f_system(N_Vector x, N_Vector f, void* user_data) {
+    const system_data* explicit_system
+        = static_cast<const system_data*>(user_data);
+    explicit_system->f_system(NV_DATA_S(x), NV_DATA_S(f));
+
+    return 0;
+  }
+
+  /**
+   * Implements the function of type CVDlsJacFn which is the user-defined
+   * callbacks for KINSOL to calculate the jacobian of the system.
+   * The Jacobian is stored in column major format.
+   *
+   * REMARK - tmp1 and tmp2 are pointers to memory allocated for variables
+   * of type N_Vector which can be used by KINJacFN (the function which
+   * computes the Jacobian) as temporary storage or work space.
+   * See
+   * https://computation.llnl.gov/sites/default/files/public/kin_guide-dev.pdf,
+   * page 55.
+   */
+  static int kinsol_jacobian(N_Vector x, N_Vector f, SUNMatrix J,
+                             void* user_data, N_Vector tmp1, N_Vector tmp2) {
+    const system_data* explicit_system
+        = static_cast<const system_data*>(user_data);
+    return explicit_system->jacobian_states(NV_DATA_S(x), J);
+  }
+
+ private:
+  /**
+   * Calculates the root function, using the user-supplied functor
+   * for a given value of x.
+   */
+  inline void f_system(const double x[], double f[]) const {
+    const std::vector<double> x_vec(x, x + N_);
+    const std::vector<double>& f_vec
+        = to_array_1d(f_(to_vector(x_vec), y_, dat_, dat_int_, msgs_));
+    std::move(f_vec.begin(), f_vec.end(), f);
+  }
```

Comment:
See the new implementation of `kinsol_f_system`. Unfortunately, you still need to make one copy of `x`, because you cannot pass an `Eigen::Map` in lieu of an `Eigen::VectorXd` for a custom function. See this [page](https://eigen.tuxfamily.org/dox/group__TutorialMapClass.html):

> All Eigen functions are written to accept Map objects just like other Eigen types. However, when writing your own functions taking Eigen types, this does not happen automatically: a Map type is not identical to its Dense equivalent.

Reply:
You can workaround this using the ```Ref<>``` class, so rather than ```foo(VectorXd)``` you use ```foo(Ref<VectorXd>)```, then ```foo``` will accept ```Map<VectorXd>```

Reply:
@andrjohns I don't think user-defined stan function takes `Eigen::Ref` type.

@charlesm93 better with constructor instead of assignment:
```c++
    Eigen::VectorXd x_eigen(Eigen::VectorXd::Map(NV_DATA_S(x), explicit_system->N_));
    Eigen::Map<Eigen::VectorXd>(N_VGetArrayPointer(f), explicit_system->N_)
      = explicit_system->f_(x_eigen, explicit_system->y_,...                       
```